File: packages/observability/src/resource.ts
/**
 * Resource attributes — the OTel "Resource" is the immutable description
 * of the entity producing telemetry. Every exporter carries it.
 *
 * Per phase-01 §5.B.3 we emit:
 *   - service.name
 *   - service.version
 *   - deployment.environment
 *   - git.sha
 *
 * Plus the optional `service.namespace` and `host.name` for multi-tenant
 * readouts in Grafana.
 *
 * The values come from a precedence chain: explicit arg > env var >
 * derived default. The explicit args let services like `apps/web` and
 * `services/realtime-gateway` override the namespace cleanly.
 */

export interface ResourceAttributes {
  'service.name': string;
  'service.version': string;
  'service.namespace'?: string;
  'deployment.environment': string;
  'git.sha': string;
  'host.name'?: string;
  [key: string]: string | undefined;
}

export interface ResourceOptions {
  serviceName: string;
  serviceVersion?: string;
  serviceNamespace?: string;
  environment?: string;
  gitSha?: string;
  hostName?: string;
  extra?: Record<string, string>;
}

const SHA_RE = /^[0-9a-f]{7,64}$/i;
const MAX_KEY_LEN = 256;
const MAX_VAL_LEN = 1024;
// Allow service.name-style identifiers: letters, digits, underscores,
// dots, and hyphens. Matches the OTel spec subset.
const SAFE_KEY_RE = /^[a-zA-Z][a-zA-Z0-9_.-]*$/;
const SAFE_VAL_RE = /^[a-zA-Z0-9._:/@=+-,]*$/;

function safeEnv(value: string | undefined, fallback: string): string {
  if (value === undefined || value.length === 0) return fallback;
  return value;
}

function readDefault(opts: ResourceOptions): ResourceAttributes {
  const serviceName = opts.serviceName;
  if (!SAFE_KEY_RE.test(serviceName) || serviceName.length > MAX_KEY_LEN) {
    throw new ResourceError(
      `service.name must match /^[a-zA-Z_][a-zA-Z0-9_.]*$/ and be <= ${MAX_KEY_LEN} chars; got ${JSON.stringify(serviceName)}`,
    );
  }

  const version = safeEnv(opts.serviceVersion, process.env['DOMIO_SERVICE_VERSION'] ?? '0.0.0+unknown');
  const env = safeEnv(opts.environment, process.env['DOMIO_ENV'] ?? process.env['NODE_ENV'] ?? 'development');
  const gitSha = safeEnv(opts.gitSha, process.env['GIT_SHA'] ?? process.env['GITHUB_SHA'] ?? 'unknown');

  if (!SHA_RE.test(gitSha) && gitSha !== 'unknown') {
    throw new ResourceError(
      `git.sha must be a hex string of 7..64 chars (or "unknown"); got ${JSON.stringify(gitSha)}`,
    );
  }

  const attrs: ResourceAttributes = {
    'service.name': serviceName,
    'service.version': version,
    'deployment.environment': env,
    'git.sha': gitSha,
  };
  if (opts.serviceNamespace) attrs['service.namespace'] = opts.serviceNamespace;
  if (opts.hostName) attrs['host.name'] = opts.hostName;
  if (opts.extra) {
    for (const [k, v] of Object.entries(opts.extra)) {
      if (!SAFE_KEY_RE.test(k) || k.length > MAX_KEY_LEN) {
        throw new ResourceError(`resource attribute key ${JSON.stringify(k)} is invalid`);
      }
      if (v.length > MAX_VAL_LEN || !SAFE_VAL_RE.test(v)) {
        throw new ResourceError(`resource attribute value for ${k} is invalid`);
      }
      attrs[k] = v;
    }
  }
  return attrs;
}

export class ResourceError extends Error {
  override readonly name = 'ResourceError';
}

/**
 * Build a `Resource` (the OTLP wire-shape subset Domio needs). The same
 * shape is reused across traces, metrics, and logs.
 */
export function buildResource(opts: ResourceOptions): ResourceAttributes {
  return readDefault(opts);
}

/**
 * Validates an OTLP HTTP endpoint string. Returns the parsed URL on
 * success, throws an `EndpointError` on failure. Empty string is NOT a
 * valid endpoint — callers should pre-check for the no-op case.
 */
export function parseOtlpEndpoint(raw: string): URL {
  if (raw.trim().length === 0) {
    throw new EndpointError('OTLP endpoint must not be empty');
  }
  let url: URL;
  try {
    url = new URL(raw);
  } catch {
    throw new EndpointError(`OTLP endpoint is not a valid URL: ${JSON.stringify(raw)}`);
  }
  const proto = url.protocol;
  if (proto !== 'http:' && proto !== 'https:') {
    throw new EndpointError(`OTLP endpoint must use http(s); got ${proto}`);
  }
  if (url.host.length === 0) {
    throw new EndpointError('OTLP endpoint is missing host');
  }
  return url;
}

export class EndpointError extends Error {
  override readonly name = 'EndpointError';
}

File: packages/observability-py/domio_obs/internal/resource.py
"""
Resource attribute builder.

Mirrors ``@domio/observability/src/resource.ts`` and the Go SDK. The
four required Phase 01 §5.B.3 attributes are always emitted; optional
attributes attach when present.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass, field
from typing import Any, Dict, Optional

MAX_KEY_LEN = 256
MAX_VAL_LEN = 1024
SAFE_KEY_RE = re.compile(r"^[a-zA-Z][a-zA-Z0-9_.\-]*$")
SAFE_VAL_RE = re.compile(r"^[a-zA-Z0-9._:/@=+,\-]*$")
SHA_RE = re.compile(r"^[0-9a-fA-F]{7,64}$")


class ResourceError(ValueError):
    """Raised when resource construction fails validation."""


@dataclass
class ResourceAttributes:
    service_name: str
    service_version: str = ""
    service_namespace: str = ""
    deployment_environment: str = ""
    git_sha: str = ""
    host_name: str = ""
    extras: Dict[str, str] = field(default_factory=dict)

    def to_otlp(self) -> Dict[str, Any]:
        pairs = [
            {"key": "service.name", "value": {"stringValue": self.service_name}},
            {"key": "service.version", "value": {"stringValue": self.service_version}},
            {"key": "deployment.environment", "value": {"stringValue": self.deployment_environment}},
            {"key": "git.sha", "value": {"stringValue": self.git_sha}},
        ]
        if self.service_namespace:
            pairs.append({"key": "service.namespace", "value": {"stringValue": self.service_namespace}})
        if self.host_name:
            pairs.append({"key": "host.name", "value": {"stringValue": self.host_name}})
        for k, v in self.extras.items():
            pairs.append({"key": k, "value": {"stringValue": v}})
        return {"attributes": pairs}


def _first_nonempty(*values: Optional[str]) -> str:
    for v in values:
        if v:
            return v
    return ""


def build_resource(
    service_name: str,
    service_version: Optional[str] = None,
    service_namespace: Optional[str] = None,
    environment: Optional[str] = None,
    git_sha: Optional[str] = None,
    host_name: Optional[str] = None,
    extra: Optional[Dict[str, str]] = None,
) -> ResourceAttributes:
    if not SAFE_KEY_RE.match(service_name) or len(service_name) > MAX_KEY_LEN:
        raise ResourceError(
            f"service.name must match ^[a-zA-Z][a-zA-Z0-9_.-]*$ and be <= {MAX_KEY_LEN} chars; "
            f"got {service_name!r}"
        )

    version = service_version if service_version is not None else os.environ.get("DOMIO_SERVICE_VERSION", "0.0.0+unknown")
    env = environment if environment is not None else _first_nonempty(
        os.environ.get("DOMIO_ENV"),
        os.environ.get("NODE_ENV"),
        "development",
    )
    sha = git_sha if git_sha is not None else _first_nonempty(
        os.environ.get("GIT_SHA"),
        os.environ.get("GITHUB_SHA"),
        "unknown",
    )

    if sha != "unknown" and not SHA_RE.match(sha):
        raise ResourceError(
            f"git.sha must be a 7..64 char hex string (or 'unknown'); got {sha!r}"
        )

    extras = dict(extra or {})
    for k, v in extras.items():
        if not SAFE_KEY_RE.match(k) or len(k) > MAX_KEY_LEN:
            raise ResourceError(f"resource attribute key {k!r} is invalid")
        if not SAFE_VAL_RE.match(v) or len(v) > MAX_VAL_LEN:
            raise ResourceError(f"resource attribute value for {k!r} is invalid")

    return ResourceAttributes(
        service_name=service_name,
        service_version=version,
        service_namespace=service_namespace or "",
        deployment_environment=env,
        git_sha=sha,
        host_name=host_name or "",
        extras=extras,
    )


# ---------- OTLP endpoint parsing ----------


class EndpointError(ValueError):
    """Raised when an OTLP endpoint URL is malformed."""


def parse_otlp_endpoint(raw: str) -> str:
    """Validates and returns the OTLP endpoint URL string."""
    if not raw or not raw.strip():
        raise EndpointError("OTLP endpoint must not be empty")
    from urllib.parse import urlparse
    try:
        u = urlparse(raw)
    except Exception as exc:
        raise EndpointError(f"OTLP endpoint is not a valid URL: {raw!r}") from exc
    if u.scheme not in ("http", "https"):
        raise EndpointError(f"OTLP endpoint must use http(s); got {u.scheme!r}")
    if not u.hostname:
        raise EndpointError("OTLP endpoint is missing host")
    return raw.rstrip("/")